azurelinuxagent/ga/collect_telemetry_events.py

```diff
@@ -248,17 +253,35 @@ def _ensure_all_events_directories_empty(extension_events_directories):
                                      ustr(error))
                         log_err = False
 
+    @staticmethod
+    def _read_event_file(event_file_path):
+        """
+        Read the event file and return the data.
+        :param event_file_path: Full path of the event file.
+        :return: Event data in list or string format.
+        """
+        # Retry for reading the event file in case file is modified while reading
+        error_count = 0
+        while True:
+            try:
+                # Read event file and decode it properly
+                with open(event_file_path, "rb") as event_file_descriptor:
+                    event_data = event_file_descriptor.read().decode("utf-8")
+
+                # Parse the string and get the list of events
+                return json.loads(event_data)
+            except Exception:
```

Comment:
this catches all kinds of errors. what is the error we were trying to handle? empty file?

Reply:
I don't remember exception type. The scenario I know retry for reading the event file in case file is modified while reading. There could be josn deserialize error or file error.

Reply:
Found FileNotFoundError and jsondecodeerror from telemetry and added those for retry

```
DroppedEventsCount: 1
Reasons (first 5 errors): [EventError] Error parsing event: not well-formed (invalid token): line 1, column 1: Traceback (most recent call last):
  File "bin/WALinuxAgent-2.11.1.12-py3.9.egg/azurelinuxagent/common/event.py", line 214, in parse_event
    return parse_json_event(data_str)
  File "bin/WALinuxAgent-2.11.1.12-py3.9.egg/azurelinuxagent/common/event.py", line 204, in parse_json_event
    data = json.loads(data_str)
  File "/usr/lib64/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python3.6/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib64/python3.6/json/decoder.py", line 355, in raw_decode
    obj, end = self.scan_once(s, idx)
json.decoder.JSONDecodeError: Expecting ',' delimiter: line 25 column 124 (char 874)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "bin/WALinuxAgent-2.11.1.12-py3.9.egg/azurelinuxagent/common/event.py", line 237, in parse_xml_event
    xml_doc = parse_doc(data_str)
  File "bin/WALinuxAgent-2.11.1.12-py3.9.egg/azurelinuxagent/common/utils/textutil.py", line 41, in parse_doc
    return minidom.parseString(xml_text)
  File "/usr/lib64/python3.6/xml/dom/minidom.py", line 1968, in parseString
    return expatbuilder.parseString(string)
  File "/usr/lib64/python3.6/xml/dom/expatbuilder.py", line 925, in parseString
    return builder.parseString(string)
  File "/usr/lib64/python3.6/xml/dom/expatbuilder.py", line 223, in parseString
    parser.Parse(string, True)
xml.parsers.expat.ExpatError: not well-formed (invalid token): line 1, column 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "bin/WALinuxAgent-2.11.1.12-py3.9.egg/azurelinuxagent/common/event.py", line 216, in parse_event
    return parse_xml_event(data_str)
  File "bin/WALinuxAgent-2.11.1.12-py3.9.egg/azurelinuxagent/common/event.py", line 247, in parse_xml_event
    raise ValueError(ustr(e))
ValueError: not well-formed (invalid token): line 1, column 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "bin/WALinuxAgent-2.11.1.12-py3.9.egg/azurelinuxagent/ga/collect_telemetry_events.py", line 437, in process_events
    event = parse_event(event_data)
  File "bin/WALinuxAgent-2.11.1.12-py3.9.egg/azurelinuxagent/common/event.py", line 218, in parse_event
    raise EventError("Error parsing event: {0}".format(ustr(e)))
azurelinuxagent.common.exception.EventError: [EventError] Error parsing event: not well-formed (invalid token): line 1, column 1
```
```
DroppedEventsCount: 583
Reasons (first 5 errors): [Errno 2] No such file or directory: '/var/lib/waagent/events/1726660508442709.tld': Traceback (most recent call last):
  File "bin/WALinuxAgent-2.11.1.12-py3.9.egg/azurelinuxagent/ga/collect_telemetry_events.py", line 434, in process_events
    with open(event_file_path, "rb") as event_fd:
FileNotFoundError: [Errno 2] No such file or directory: '/var/lib/waagent/events/1726660508442709.tld'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "bin/WALinuxAgent-2.11.1.12-py3.9.egg/azurelinuxagent/ga/collect_telemetry_events.py", line 461, in process_events
    os.remove(event_file_path)
FileNotFoundError: [Errno 2] No such file or directory: '/var/lib/waagent/events/1726660508442709.tld'
, [Errno 2] No such file or directory: '/var/lib/waagent/events/1726660507946074.tld': Traceback (most recent call last):
  File "bin/WALinuxAgent-2.11.1.12-py3.9.egg/azurelinuxagent/ga/collect_telemetry_events.py", line 434, in process_events
    with open(event_file_path, "rb") as event_fd:
FileNotFoundError: [Errno 2] No such file or directory: '/var/lib/waagent/events/1726660507946074.tld'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "bin/WALinuxAgent-2.11.1.12-py3.9.egg/azurelinuxagent/ga/collect_telemetry_events.py", line 461, in process_events
    os.remove(event_file_path)
FileNotFoundError: [Errno 2] No such file or directory: '/var/lib/waagent/events/1726660507946074.tld'
, [Errno 2] No such file or directory: '/var/lib/waagent/events/1726660507845022.tld': Traceback (most recent call last):
  File "bin/WALinuxAgent-2.11.1.12-py3.9.egg/azurelinuxagent/ga/collect_telemetry_events.py", line 434, in process_events
    with open(event_file_path, "rb") as event_fd:
FileNotFoundError: [Errno 2] No such file or directory: '/var/lib/waagent/events/1726660507845022.tld'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "bin/WALinuxAgent-2.11.1.12-py3.9.egg/azurelinuxagent/ga/collect_telemetry_events.py", line 461, in process_events
    os.remove(event_file_path)
FileNotFoundError: [Errno 2] No such file or directory: '/var/lib/waagent/events/1726660507845022.tld'
, [Errno 2] No such file or directory: '/var/lib/waagent/events/1726660508044223.tld': Traceback (most recent call last):
  File "bin/WALinuxAgent-2.11.1.12-py3.9.egg/azurelinuxagent/ga/collect_telemetry_events.py", line 434, in process_events
    with open(event_file_path, "rb") as event_fd:
FileNotFoundError: [Errno 2] No such file or directory: '/var/lib/waagent/events/1726660508044223.tld'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "bin/WALinuxAgent-2.11.1.12-py3.9.egg/azurelinuxagent/ga/collect_telemetry_events.py", line 461, in process_events
    os.remove(event_file_path)
FileNotFoundError: [Errno 2] No such file or directory: '/var/lib/waagent/events/1726660508044223.tld'
, [Errno 2] No such file or directory: '/var/lib/waagent/events/1726660502444669.tld': Traceback (most recent call last):
  File "bin/WALinuxAgent-2.11.1.12-py3.9.egg/azurelinuxagent/ga/collect_telemetry_events.py", line 434, in process_events
    with open(event_file_path, "rb") as event_fd:
FileNotFoundError: [Errno 2] No such file or directory: '/var/lib/waagent/events/1726660502444669.tld'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "bin/WALinuxAgent-2.11.1.12-py3.9.egg/azurelinuxagent/ga/collect_telemetry_events.py", line 461, in process_events
    os.remove(event_file_path)
FileNotFoundError: [Errno 2] No such file or directory: '/var/lib/waagent/events/1726660502444669.tld'
```